ocaml/xapi/xapi_vgpu_type.ml

```diff
@@ -421,19 +421,52 @@ module Vendor_nvidia = struct
         !Xapi_globs.nvidia_host_driver_file ;
       Xapi_globs.nvidia_default_host_driver_version
 
+  (** [compare_versions] compares two strings representing host driver
+  versions. These are numbers separated by dots. It may also include a
+  'a' which is discarded. This is not a complete syntax validation. *)
+  let compare_version v1 v2 =
+    let of_string version =
+      try
+        Astring.String.filter (( <> ) '*') version
+        |> Astring.String.cuts ~empty:false ~sep:"."
+        |> List.map int_of_string
+      with _ ->
+        error "Not a proper NVidia host driver version: '%s'" version ;
+        failwith (Printf.sprintf "%s is not a version string" version)
+    in
+    let rec compare v1 v2 =
+      match (v1, v2) with
+      | [], [] ->
+          0
+      | [], 0 :: ys ->
+          compare v1 ys
+      | [], _ ->
+          -1
+      | 0 :: xs, [] ->
+          compare xs v2
+      | _, [] ->
+          1
+      | x :: xs, y :: ys when x < y ->
+          -1
+      | x :: xs, y :: ys when x > y ->
+          1
+      | _ :: xs, _ :: ys ->
+          compare xs ys
+    in
+    compare (of_string v1) (of_string v2)
+
   let host_driver_supports_multi_vgpu ~host_driver_version ~supported_versions =
-    let version = float_of_string host_driver_version in
-    List.exists
-      (fun pattern ->
-        match String.index_opt pattern '+' with
-        | Some i -> (
-          try float_of_string (String.sub pattern 0 i) <= version
-          with _ -> false
-        )
-        | None -> (
-          try float_of_string pattern = version with _ -> false
-        ))
-      supported_versions
+    try
+      let latest =
+        List.find_opt (fun v -> String.contains v '*') supported_versions
```

Comment:
It used to be `+` rather than `*`?

Reply:
Yeah - and the unit tests found it, too. Fixed now. I have tested this otherwise manually.

Reply:
There is another test failure - have to review it